### PR TITLE
QMK CLI and JSON keymap support

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

This PR adds fleshed out python support to QMK. I tried to keep the changes fairly minimal to start, but it does comprise several features:

* Fleshed out python support- it should now be easier for people to add python tools to qmk
* QMK CLI (Add `qmk_firmware/bin` to your path to use
* JSON keymap support- you can now use `keymap.json` instead of `keymap.c`.
* `qmk compile-json` - a command that will compile configurator exports
* `qmk hello` - an example QMK CLI command

## Types of Changes

- [x] Core
- [x] New feature
- [x] Keymap/layout/userspace (addition or update)

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: build_json.mk [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Look for a json keymap file
+ifneq ("$(wildcard $(MAIN_KEYMAP_PATH_5)/keymap.json)","")
```
Comment (yanfali) [inline build_json.mk]: can you use a loop against a these wildcard paths?

**Comment by yanfali** [comment]
OK, it looks like it only works from the top level. I don't really like that assumption. Anyway you can make it work from any dir? it's not make, it's python, you should be able to find the basedir easily.

**Comment by yanfali** [comment]
The usage messages are too terse. A couple of lines about what they do or pointing to documentation would be helpful. I can't figure out what qmk-json-keymap does for example. I tried the same command as compile and it just errored out. There's no obvious clue as to what it actually does.

**Comment by yanfali** [comment]
```shell
qmk-json-keymap keyboards/clueboard/66_hotswap/keymaps/json/keymap.json
☒ 'Namespace' object has no attribute 'output'
Traceback (most recent call last):
  File "/Users/yanfa.li/projects/qmk_firmware/lib/python/milc.py", line 537, in __call__
    self.__call__()
  File "/Users/yanfa.li/projects/qmk_firmware/lib/python/milc.py", line 543, in __call__
    return self._entrypoint(self)
  File "/Users/yanfa.li/projects/qmk_firmware/lib/python/qmk/cli/json/keymap.py", line 25, in main
    if cli.args.output == ('-'):
AttributeError: 'Namespace' object has no attribute 'output'
➜  qmk_firmware git:(compile_json) ✗ qmk-json-keymap                        <<<
usage: qmk-json-keymap [-h] [-V] [-v] [--datetime-fmt GENERAL_DATETIME_FMT]
                       [--log-fmt GENERAL_LOG_FMT]
                       [--log-file-fmt GENERAL_LOG_FILE_FMT]
                       [--log-file GENERAL_LOG_FILE] [--color] [--no-color]
                       [-c GENERAL_CONFIG_FILE] [--save-config]
                       [-o GENERAL_OUTPUT]
                       filename
qmk-json-keymap: error: the following arguments are required: filename
```

File: build_json.mk (hunk 1 above)
Comment (skullydazed) [reply build_json.mk]: Make is a pretty limited language. This is the pattern we use several other places.

**Comment by skullydazed** [comment]
I fixed up some of the file handling and made the CLI a little more robust in the face of errors. Bugs identified so far should be addressed.

**Comment by gwerbin** [comment]
What's the intended installation process for this?

If you're already asking users to `pip install -r requirements.txt`, then you might as well go the extra mile and make the whole thing an installable package. That has the added advantage of setting this code up to become a general-purpose QMK utility library.

I'm experienced with Python application development and packaging, so if you're interested in what I describe below I am happy to contribute the time to making it happen.

Right now you are "installing" scripts into `bin/` and manually dispatching thereto from `bin/qmk`. That certainly works for this particular, case, but I have two specific recommendations:

1. Use the built-in `argparse` library, which provides the ability to parse `-options` and `--options` as well as delegate to subcommands
2. Make the whole thing "installable", *and* use "entry points" so that the command line program `qmk` will be accessible from anywhere, as long as Python where it was installed is on the user's PATH

Recommendation 1 means no more manual parsing of `sys.argv` -- it's optional but nice to have. It also means automatically-generated `--help` text, automatic/default error messages, et al.

Recommendation 2 means you can delete the `bin/qmk*` files entirely, and replace them with a `__main__.py` file that implements the whole CLI in pure Python functions, and you never have to worry about the shebang line being incorrect.

If you're interested I can put together a PoC based off this PR.

**Comment by e11i0t23** [comment]
If you where to implement recommendation 2 of gwerbin's recommendations. I would recommend adding a way to set qmk dir for global usage. As well as possibly the ability to clone the repo. Issue with cloning is we should be using forks whereas this would have to clone master really.

**Comment by skullydazed** [comment]
> What's the intended installation process for this?

When users set up their build environment using `util/qmk_install.sh` it installs python3 and then runs `pip3 -r requirements.txt`. After that `bin/qmk` can be run from any directory (and/or put into your path if you desire.)

In the past we had a goal of producing a package that could be on pypi (and maybe brew/pacman/etc) but for the time being we've decided we need to move too quickly for that to be practical. With python in QMK we're about to enter a period of heavy iteration, so having to coordinate changes to both python and the build system would be a big hinderance.

> Make the whole thing "installable", and use "entry points" so that the command line program qmk will be accessible from anywhere, as long as Python where it was installed is on the user's PATH

Entrypoints are on my radar, and something I plan to use in the future. If you'd like to take on that work you're welcome to. As I said in the last paragraph an installable CLI is an anti-goal for the short term, but that's not required to benefit from entrypoints.

**Comment by awkannan** [comment]
Is there documentation anywhere about what this CLI is supposed to achieve? Looking in the PR it's still unclear to me. `The QMK CLI makes building and working with QMK keyboards easier` is pretty vague, do we have specific examples of things that this would help with other than compile-json? Or is it meant to be vague?

What does compiling a json mean? Does it make your keyboard.h LAYOUT macro and a default layout for you? Does it take a JSON file and compile an actual keyboard firmware for it using QMK? It would be good to call that out. 

An installable CLI is only useful if you also have the QMK repo already setup, right? I think that should also be documented. I could imagine someone new pulling down the repo, then using bin/qmk and complaining that it doesn't work.

In general, I don't think there is any issue with the code, but I think the documentation could be greatly fleshed out. I have no idea what would/wouldn't be accepted as a CLI command as it stands, and I think this could cause a lot of confusion if not rectified.

**Comment by skullydazed** [comment]
> Or is it meant to be vague?

It's vague because we don't have a lot of functionality defined yet. This is still early in the process and we don't want to box things in too much yet. There will be other CLI scripts for things that are obvious, EG new keyboard and new keymap, but I also want to leave room for things that aren't obvious yet. We also don't want to over promise. :)

> What does compiling a json mean? 

This is spelled out in cli.md:

```
This command allows you to compile JSON files you have downloaded from <https://config.qmk.fm>.
```

> I could imagine someone new pulling down the repo, then using bin/qmk and complaining that it doesn't work.

This is a good point. We should make sure it fails gracefully if modules aren't installed. We have also talked about having `qmk doctor` for investigating environment issues in general, but so far no one has taken that on.

> I have no idea what would/wouldn't be accepted as a CLI command as it stands

That makes two of us. ;)

To your point, some of what you complain about is a deliberate choice. We are still working out the right decisions and so we don't want to constrain prematurely. These things will shake out soon enough, but in the meantime cementing a bad decision has consequences that are hard to undo later.

**Comment by yanfali** [comment]
Yeah qmk doctor is on my radar. On Linux we'd like it to detect modem manager, and recommend udev rules. On windows detect drivers are missing for dfu. On all platforms that required binaries are in the CLI path.

**Comment by yanfali** [comment]
```python3
#!/usr/local/bin/python3

import shutil

binaries = [
    'dfu-programmer', 'avrdude', 'dfu-util', 'avr-gcc', 'arm-none-eabi-gcc'
]

print('QMK Doctor is Checking your environment')

ok = True
for binary in binaries:
    res = shutil.which(binary)
    if res is None:
        print('QMK can\'t find ' + binary + ' in your path')
        ok = False

if ok:
    print('QMK is ready to go')
```

Here's my naive start to a doctor command. It does almost nothing, but it would be helpful to run a few tests before people ask for help on qmk discord. A lot of the questions tend to be very similar, so we could save some time if people could help themselves.

**Comment by drashna** [comment]
I know it's adding more and more, but for the qmk-doctor, would it be worth checking the AVR and ARM binary versions, too?

**Comment by yanfali** [comment]
> I know it's adding more and more, but for the qmk-doctor, would it be worth checking the AVR and ARM binary versions, too?

Yeah, it's pretty easy. Just need to parse the response of the command. With system I'm only checking the return code. You could give it a try Drashna, I think something like system.popen, whatever can give you stdout or stderr, then you can do some simple matching. Or you could do something dumb like grepping over a pipe. I you want to do math comparisons like greater than or less than though, you'll need to parse the actual output.

**Comment by skullydazed** [comment]
Having `qmk doctor` do more is a great idea, but we shouldn't let it hold up this PR. There are other PR's that will be affected by this one (particularly #6066) and I'd like to see us get this without expanding the scope too much.

File: docs/coding_conventions_python.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+```
+FIXME(username): Revisit this code when the frob feature is done.
+```
+
+...where username is your GitHub username.
+
+# Unit Tests
+
+These are good. We should have some one day.
```
Comment (yanfali) [inline docs/coding_conventions_python.md]:
```suggestion
These are good. We should have some one day.

# Editor Support

## vim

If you use w0rp/ale you can easily add linting and formatting to your `.vimrc` by installing flake8 and yapf, and adding the following lines:

```vimrc
let g:ale_linters = {
\   'python': ['flake8'],
\}
let g:ale_fixers = {
      \ 'python': ['yapf']
\}
```
```

**Comment by orngshdw** [comment]
Hi sorry for the late comment. 
It may be worthwhile to consider not installing (or ask before installing) python packages in requirements.txt globally as it may conflict with people who version control packages installed on their machine (since requirements.txt doesn't specify versions, the latest of each module will be installed). 

One suggestion is to use a python virtual environment, such as [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), to install packages in and run our python scripts with. If we do that, we can avoid the issue I mentioned above & have the added benefit of knowing that the python scripts are being run with the modules & versions we specify in the requirements.txt file. It can be something as simple as:
```bash
# creating the python3 virutalenv in qmk project root
python3 -m venv qmk_pythonenv

# activate the python environment created
. qmk_pythonenv/bin/activate

# all pip/python commands will install in that virtualenv
which python
>>> ${qmk_firmware_root}/qmk_pythonenv/bin/python
which pip
>>> ${qmk_firmware_root}/qmk_pythonenv/bin/python
####
# run qmk commands or install modules here
####

# deactivate virtual enviroment at end of script/work
deactivate

# all python/pip commands after the deactivate command now use global python/pip environment
``` 

Note we'll have to add a directory to the .gitignore file to ignore the virtualenvironment directory we create.

**Comment by yanfali** [comment]
@orngshdw  good points, but this just makes everything even harder to use. The opposite of the intention of the command :(

**Comment by awkannan** [comment]
@yanfali @orngshdw as someone who works in python daily, PLEASE use a virtualenv! The fact that the install is hidden inside a shell script makes it even more insidious.

My recommendation is to use [pipenv](https://github.com/pypa/pipenv) to make it dirt simple and easy.

**Comment by skullydazed** [comment]
I was hoping to avoid virtualenv at first. :)

You make good points on how we can do what we need without impacting users. The question is whether we should let that work hold up getting this in.

I'd like to note that we install a lot of software during the install process, some of which may conflict with a user's needs. The risk of adding the two packages we currently do is minimal, and if someone needs to pin those to another version it likely won't impact us for now. That's why I've left only abstract package definitions for the moment. We can add a prompt and/or a note to the documentation so that people are aware, but I don't think that particular problem is python specific.

The right way to use virtualenv here is to create the environment and use virtualenv entrypoints so that `bin/qmk` works as expected without having to activate. As @yanfali says we don't want to increase complexity for the user, and luckily we don't have to. We can create the virtualenv and the proper symlinks during the setup process.

My experience with virtualenv is that getting to that point will take some elbow grease. That will get done faster if someone volunteers to take that project on.

**Comment by yanfali** [comment]
@awkannan not disagreeing, this is more of a case of python packaging sucking

**Comment by yanfali** [comment]
So in reality this only really affects python developers. Most end users probably don't care about polluting their global python cache. Does it make sense to use something like pipenv to enforce the python environment if we detect it on the runtime path?

**Comment by gwerbin** [comment]
I strongly, strongly recommend the virtualenv approach. It's just too easy to break a Python installation.  I'd estimate that 10 to 20% of all newbie questions I field on chat rooms are related to broken environments. And heaven forbid someone should decide to become a python developer after they have QMK installed. And then what?

```bash
python -m venv ./venv
. ./venv/bin/activate
python -m pip install requirements.txt
```
Is that really so bad for first time set up?

And similarly is
```bash
. ./venv/bin/activate
```
Before running QMK commands that bad?

Take it from the experienced python users here, the alternatives are painful and they are not worth the pain.

**Comment by yanfali** [comment]
@gwerbin have you seen most of our users, they have a hard enough time installing drivers. Asking them to jump through venv; which incidentally requires more software to install, is even more pain and raises barriers for entry. In principle I'm not disagreeing, it's just python doesn't make this easy at all.

**Comment by skullydazed** [comment]
20 year python veteran here- yes it's the better way, and long term the way we're going to do it, but please don't oversell the problem and downplay the work involved in the solution.

Anything extra the user has to do is that bad. We're working towards a 1 command QMK install, and forcing the user to interact with virtualenv in that way prevents that. Luckily that's not the only way to use virtualenv.

Things are not going to be perfect on day 1, but we will keep improving them.

**Comment by gwerbin** [comment]
venv ships with Python. And compared to drivers it's relatively easy to debug,  not reliant on unmaintained and closed source software, generally less mysterious to all parties involved, and in an absolute worst case scenario you can just uninstall and start again.

Compared to the overhead of installing Python itself, spitting up an environment I don't think it's a big deal.

**Comment by skullydazed** [comment]
Does that mean you're volunteer to do the venv integration work here?

**Comment by orngshdw** [comment]
Think my earlier comment may have suggested something a bit unrealistic and/or had high effort to implement; hopefully you all won't view it as the only viable solution.

This PR does looks like a good first step and isn't worth the holdup as a final solution is being looked into as @skullydazed has said. I don't feel the two packages (argcomplete & colorama) are likely to be dependancies that will break other tools and are lightweight enough to uninstall via `pip uninstall -r requirements.txt` (for those of us in the know).  Others who don't use python only want qmk working so that their keyboard can be flashed afterall...

@gwerbin Hopefully anyone that becomes a python dev after installing qmk will know to only do python work in a virtualenv :) In new virtualenvs `pip freeze` doesn't return anything regardless of what is installed globally. lmk if I'm wrong

Note: there are other less impactful & simpler alternatives that will not need any user interaction (such as adding packages to the repo ourselves and then doing a `sys.path.insert`/`append` during runtime). It'll take some looking into & test for sure, which I think we can look into perhaps in another PR.

**Comment by gwerbin** [comment]
@orngshdw  I actually didn't know you could uninstall like that. I will shut up now :)

Yes this seems like a great start.

**Pull request merged**